{% extends 'layout.html' %}
{% block body %}
    <div class="panel-group">
      <h3>Tutorials</h3>
      <ul class="nav nav-tabs help-tabs" role="tablist">
        <li class="active"><a href="#regulator-centric" role="tab" data-toggle="tab">Regulator-centric</a></li>
        <li><a href="#custom-regulator" role="tab" data-toggle="tab">Custom regulator</a></li>
      </ul>

      <div class="tab-content">
        <div class="tab-pane active" id="regulator-centric">
          <h4>Example: Mutual exclusive Srsf1 and Srsf2 binding on exons</h4>

          <p>SR proteins are RNA-binding proteins that are associated with alternative splicing,
          and recent findings suggest that they are involved in both exon skipping and
          inclusion events (<a href="http://doi.org/10.1016/j.molcel.2013.:w03.001">Pandit <em>et al.</em> 2013</a>).
          Furthermore, Pandit <em>et al.</em> suggest although Srsf1 and Srsf2 binding sites
          largely target the same exons, certain exons are regulated by only one of the two
          factors, such that there is mutually exclusive binding of Srsf1 or Srsf2. Exactly
          these exons can be identified with doRiNA 2 in the following manner:</p>

          <p>As we're not uploading a custom regulator, we just click the blue "Search" button
          on the left side of the start page.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/start_page_search.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              doRiNA 2 start page, click "Search".
            </div>
          </div>
          </p>

          <p>Having clicked "Search", we now can select the organism and assembly we want to
          work with. We select the mouse assembly "mm9", and click the "Search Database" button.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/select_assembly_mm9.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Select the mm9 assembly and click "Search Database".
            </div>
          </div>
          </p>

          <p>In the search page, we can build up the search we are interested in. We don't want to
          limit our search to a set of candidate genes, so we ignore that option. As we are only
          interested in exons, we select "CDS" in the "Look for binding sites in region" settings.
          Then we use the autocompletion option of the "Regulators (set A)" field to quickly find
          the Srsf1 HITS-CLIP dataset by typing "Srsf1" and hitting the return key. Alternatively,
          we could have used the mouse to select the entry, or even just scrolled the list without
          narrowing down the displayed regulators.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/search_set_a_srsf1.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              <ol>
                <li>Look for binding sites in CDS regions only</li>
                <li>Type "Srsf1" into the "Regulators (set A)" field to search for
                the Srsf1 HITS-CLIP dataset</li>
              </ol>
            </div>
          </div>
          </p>

          <p>Now to repeat the process for Srsf2, we first expand the combinatorial search settings
          by clicking on "Combinatorial search". Then we again look for binding sites in CDS regions
          and pick the Srsf2 HITS-CLIP dataset for "Regulators (set B)". As we are interested in
          mutually exclusive binding, we use the "XOR" (exclusive or) set operation. Then, we click
          the blue "Search doRiNA" button to start the search.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/search_set_b_srsf2.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              <ol>
                <li>Expand the combinatorial search pane</li>
                <li>Look for binding sites in CDS regions only</li>
                <li>Type "Srsf2" into the "Regulators (set B)" field to search for
                the Srsf1 HITS-CLIP dataset</li>
                <li>Select XOR to search for mutually exclusive binding</li>
                <li>Click the blue "Search doRiNA" button to start the search</li>
              </ol>
            </div>
          </div>
          </p>

          <p>After a few seconds, the results page is displayed, giving 44,476 results that are
          bound by either Srsf1 or Srsf2, but not both. Buttons at the top and the bottom of the
          page allow you to start a new search (chose this if you want to search on a different
          assembly) or to go back to the search options to redo the search on the same assembly
          with changed search options. A button on the top of the page allows you to download the
          complete results in BED file format. To optimise loading times, the results page only
          displays the top 100 hits, a button on the bottom of the page allows you to load more
          results.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/results_page_srsf1_2.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Results page, abridged.
            </div>
          </div>
          </p>
        </div>

        <div class="tab-pane" id="custom-regulator">
          <h4>Example: Genic regions that form circular RNAs and overlap with Ago sites in
              HEK293 cells</h4>
          <p>Circular RNAs have emerged as a new regulatory RNA species in recent years
          (<a href="http://doi.org/10.1126/science.1238522">Wilusz &amp; Sharp, 2013</a>).
          The discovery that circRNAs act as "sponges" for miRNA highlights one of the potentially
          many regulatory roles of circRNAs. circRNAs are typically generated from primary
          transcripts that also give rise to linear transcript isoforms. Using a circRNA, we will
          show how to use the custom file upload. The
          <a href="http://www.circbase.org/download/hsa_hg19_Memczak2013_HEK293.bed">BED file of all
          detected circles in HEK293</a> from the study of
          <a href="http://doi.org/10.1038/nature11928">Memczak et al 2013</a>
          was downloaded from <a href="http://www.circbase.org/">circBase</a>.</p>

          <p>We want to upload the HEK293 circRNA BED file from above, so we select "upload" this
          time around. This opens your browser's file dialog, browse and select the downloaded BED
          file.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/start_page_upload.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Start page, select "upload" to pick a custom regulator BED file to upload.
            </div>
          </div>
          </p>

          <p>The start page changes to reflect that a file was selected, now click "Go" to upload
          the file and start the search.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/start_page_upload_selected.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Start page, custom regulator BED file selected.
            </div>
          </div>
          </p>

          <p>We can now select the organism and assembly we want to
          work with. We select the human assembly "hg19", and click the "Search Database" button.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/select_assembly_hg19.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Select the hg19 assembly and click "Search Database".
            </div>
          </div>
          </p>

          <p>On the search configuration page, the "uploaded custom regulator" is already pre-selected in
          the "Regulators (set A) field. Using the interactive search, we type "ago2" to quickly find the
          AGO2 CLIP-Seq regulator. Once we have selected that, we can decide if we want to require a match
          to all of the multiple regulators for a hit or if a match to any regulator is sufficient. As we
          are going to restrict our hits to circRNAs that are no more than 100 bp away from the RBP targets
          in the "Restrict to overlapping features with a maximal distance" field, we can leave the "Regulators
          that must match" on "any".<br>
          Expanding the "Combinatorial search" pane again, we select the "Regulators (set B)" field, and there
          type "custom" to find our custom regulator. We then type "ago1234" to find the AGO1234 PAR-CLIP
          regulator dataset. We can again leave "Regulators that must match" on "any", and we can also leave
          "Define set operation" on "OR" to get the union of all results. Last but not least we again "Restrict
          to overlapping features with a maximal distance" of 100 bp. Once that is done, we can click "Search
          doRiNA" to start the search.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/search_custom_ago2_ago1234.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              <ol>
                <li>Keep "Regulators that must match" on "any"</li>
                <li>"uploaded custom regulator" is already selected, type "ago2" in the
                "Regulators (set A)" field to quickly select the AGO2 CLIP-Seq dataset</li>
                <li>Limit search to circRNAs less than 100 bp away from AGO2 target sites</li>
                <li>Expand the combinatorial search</li>
                <li>Keep "Regulators that must match" on "any"</li>
                <li>In "Regulators (set B), first type "custom" to find the uploaded custom regulator,
                then type "ago1234" to find the AGO1234 PAR-CLIP dataset</li>
                <li>Keep "Define set operations" on OR to find the union of all the results</li>
                <li>Limit search to circRNAs less than 100 bp away from AGO1234 target sites</li>
                <li>Hit the "Search doRiNA" button to start the search</li>
              </ol>
            </div>
          </div>
          </p>

          <p>After a few seconds, the results page is displayed, giving 4,618 results. As we
          requested the union of all results, we get a couple of hits where either circRNAs
          overlap with other circRNAs or AGO sites overlap with each other. We can use the
          download button at the top of the page to download the full results in BED file format,
          and can then find after further analysis that 889 circRNAs are less than 100 bp away from
          566 PAR-CLIP and 606 CLIP-Seq AGO target sites.<br>
          Buttons at the top of the page allow you to start a new search discarding your uploaded
          regulator BED file or to redo the search, keeping your uploaded file around until that search
          is completed. To optimise loading times, the results page only
          displays the top 100 hits, a button on the bottom of the page allows you to load more
          results.
          <div class="thumbnail">
            <img src="{{ url_for('static', filename='images/results_page_custom.png') }}" width="66%" class="img-responsive">
            <div class="caption">
              Results page, abridged.
            </div>
          </div>
          </p>
        </div>
        </div>
      </div>
    </div>
{% endblock %}
{% block extra_scripts %}
    <script type="text/javascript">
      $(document).ready(function() {
        $('.help-tabs a[href=' + $(location).attr('hash') + ']').tab('show')
      });
    </script>
{% endblock %}
